### PR TITLE
don't save copies FileInfo

### PR DESCRIPTION
**Pull request by userz58**
use hash column to find previously uploaded files.

### CODE REVIEW

**Comment by grena** [comment]
Hello @userz58, thank you very much for contributing to Akeneo :rocket: 

Please add this line in the [CHANGELOG.md file](https://github.com/akeneo/pim-community-dev/blob/master/CHANGELOG-1.6.md) to document the BC break introduced by your PR:
```
Change constructor of `Akeneo\Component\FileStorage\File\FileStorer`. Add `Akeneo\Bundle\FileStorageBundle\Doctrine\ORM\Repository\FileInfoRepository`.
```

Here are the next steps before having your code on the PIM:

1. Here is [the result of the Travis build](https://travis-ci.org/akeneo/pim-community-dev/jobs/113289205), your code breaks some PHPspec tests, but it's ok! :) Please fix them and add some for the new methods added. If you are struggling on writing phpspec tests, we can help you to write them :muscle:
2. The core team will review your Pull Request
3. We'll run integration tests to see if it doesn't break anything in the whole app
4. Then you'll can merge!

Thanks again :)

File: src/Akeneo/Component/FileStorage/File/FileStorer.php [hunk 1]
```diff
@@ -31,19 +32,25 @@ class FileStorer implements FileStorerInterface
...
      * @param SaverInterface           $saver
      * @param FileInfoFactoryInterface $factory
+     * @param FileInfoRepository       $repository
      */
     public function __construct(
         MountManager $mountManager,
         SaverInterface $saver,
-        FileInfoFactoryInterface $factory
+        FileInfoFactoryInterface $factory,
+        FileInfoRepository $repository
```
Comment (grena) [inline src/Akeneo/Component/FileStorage/File/FileStorer.php]: Please use the `FileInfoRepositoryInterface`

File: src/Akeneo/Bundle/FileStorageBundle/Doctrine/ORM/Repository/FileInfoRepository.php [hunk 2]
```diff
@@ -29,4 +29,10 @@ public function findOneByIdentifier($identifier)
     {
         return $this->findOneBy(['key' => $identifier]);
     }
+
```
Comment (grena) [inline src/Akeneo/Bundle/FileStorageBundle/Doctrine/ORM/Repository/FileInfoRepository.php]: Extra space, you can drop it

File: src/Akeneo/Bundle/FileStorageBundle/Resources/config/file_storage.yml [hunk 3]
```diff
@@ -11,6 +11,8 @@ services:
             - '@oneup_flysystem.mount_manager'
             - '@akeneo_file_storage.saver.file'
             - '@akeneo_file_storage.file_storage.file_info_factory'
+            - '@akeneo_file_storage.repository.file_info'
+
```
Comment (grena) [inline src/Akeneo/Bundle/FileStorageBundle/Resources/config/file_storage.yml]: Extra space, you can drop it

File: src/Akeneo/Component/FileStorage/File/FileStorer.php [hunk 4]
```diff
@@ -81,13 +116,10 @@ public function store(\SplFileInfo $localFile, $destFsAlias, $deleteRawFile = fa
...
         $this->saver->save($file);
 
-        if (true === $deleteRawFile) {
-            $this->deleteRawFile($localFile);
-        }
-
         return $file;
     }
 
+
```
Comment (grena) [inline src/Akeneo/Component/FileStorage/File/FileStorer.php]: Extra space, you can drop it

File: src/Akeneo/Component/FileStorage/Repository/FileInfoRepositoryInterface.php [hunk 5]
```diff
@@ -14,4 +14,5 @@
  */
 interface FileInfoRepositoryInterface extends ObjectRepository, IdentifiableObjectRepositoryInterface
 {
+    public function findOneByHash($hash);
```
Comment (grena) [inline src/Akeneo/Component/FileStorage/Repository/FileInfoRepositoryInterface.php]: Please add phpdoc for this method

File: src/Akeneo/Bundle/FileStorageBundle/Doctrine/ORM/Repository/FileInfoRepository.php [hunk 6]
```diff
@@ -29,4 +29,10 @@ public function findOneByIdentifier($identifier)
...
+
+    public function findOneByHash($hash)
```
Comment (grena) [inline src/Akeneo/Bundle/FileStorageBundle/Doctrine/ORM/Repository/FileInfoRepository.php]: Please add inherited phpdoc for this method
```
    /**
     * {@inheritdoc}
     */
    public function findOneByHash($hash)
```

File: src/Akeneo/Component/FileStorage/File/FileStorer.php [hunk 7]
```diff
@@ -31,19 +32,25 @@ class FileStorer implements FileStorerInterface
     /** @var FileInfoFactoryInterface */
     protected $factory;
 
+    /** @var  FileInfoRepository */
```
Comment (grena) [inline src/Akeneo/Component/FileStorage/File/FileStorer.php]: Extra space after `@var` (and don't forget to update to `FileInfoRepositoryInterface` :+1: )

File: src/Akeneo/Component/FileStorage/File/FileStorer.php [hunk 8]
```diff
@@ -31,19 +32,25 @@ class FileStorer implements FileStorerInterface
...
+    protected $repository;
+
     /**
      * @param MountManager             $mountManager
      * @param SaverInterface           $saver
      * @param FileInfoFactoryInterface $factory
+     * @param FileInfoRepository       $repository
```
Comment (grena) [inline src/Akeneo/Component/FileStorage/File/FileStorer.php]: `FileInfoRepositoryInterface`

File: src/Akeneo/Component/FileStorage/File/FileStorer.php [hunk 9]
```diff
@@ -51,6 +58,34 @@ public function __construct(
...
+        return $file;
+    }
+
+    /**
+     * @param \SplFileInfo $file
+     */
+    public function findCopy(\SplFileInfo $localFile)
+    {
+        $hash = sha1_file($localFile->getPathname());
+        return $this->repository->findOneByHash($hash);
```
Comment (phaseinducer) [inline src/Akeneo/Component/FileStorage/File/FileStorer.php]: Blank line before return please ; )

File: src/Akeneo/Component/FileStorage/File/FileStorer.php [hunk 10]
```diff
@@ -51,6 +58,34 @@ public function __construct(
...
+
+        if (true === $deleteRawFile) {
+            $this->deleteRawFile($localFile);
+        }
+
+        return $file;
+    }
+
+    /**
+     * @param \SplFileInfo $file
```
Comment (phaseinducer) [inline src/Akeneo/Component/FileStorage/File/FileStorer.php]: missing `@return`

File: src/Akeneo/Component/FileStorage/File/FileStorer.php [hunk 11]
```diff
@@ -51,6 +58,34 @@ public function __construct(
...
+     */
+    public function findCopy(\SplFileInfo $localFile)
```
Comment (phaseinducer) [inline src/Akeneo/Component/FileStorage/File/FileStorer.php]: Should be protected imho

File: src/Akeneo/Component/FileStorage/File/FileStorer.php [hunk 12]
```diff
@@ -51,6 +58,34 @@ public function __construct(
...
+    {
+        $hash = sha1_file($localFile->getPathname());
+        return $this->repository->findOneByHash($hash);
+    }
+
+    /**
+     * @param \SplFileInfo $file
+     * @param string $destFsAlias
+     */
+    public function saveFile(\SplFileInfo $localFile, $destFsAlias)
```
Comment (phaseinducer) [inline src/Akeneo/Component/FileStorage/File/FileStorer.php]: idem, protected

**Comment by userz58** [comment]
#3572
scr

**Comment by juliensnz** [comment]
GTM, awesome PR

**Comment by solivier** [comment]
gtm

**Comment by skeleton** [comment]
GTM !

**Comment by willy-ahva** [comment]
Hi @userz58,

Thank you for contributing :) This PR is very nice !
Just one thing before my GTM, PHPSpecs need to be updated.

Don't hesitate to ask help if you need it.

**Comment by a2xchip** [comment]
Is it used for detection of already uploaded files?

**Comment by userz58** [comment]
Hi @willy-ahva,
I'm sorry, i do not know how to write tests (

**Comment by jjanvier** [comment]
Hello @userz58 

Don't worry, I fixed the tests for you here https://github.com/akeneo/pim-community-dev/pull/4337 and rebased the branch on master.
Your contribution will be merged once the CI will have run.

Thanks you very much :)

**Pull request closed without merging**